### ISSUE TITLE
Cannot start vnc-recorder container

### ISSUE THREAD

**Issue by cchanning** [testcontainers/testcontainers-java#308]
Hi,
I'm in the process of moving a local build (currently on Mac) out to Jenkins (running Ubuntu 16.10 + Docker 1.13.1). With Jenkins I use docker containers for build agents ("child" containers are actually created as siblings on the docker host so they're not nested). However, I'm getting the following error when trying to start a BrowserWebDriverContainer. On my local build (running Docker for Mac [latest]) I'm not getting this issue. Could you shed some insight on what the cause might be and if there are any workarounds please? For the Jenkins build, I'm currently having to skip screen recording which isn't ideal.

LOG:
12:01:26 12:01:26.559 [main] ERROR docker[richnorth/vnc-recorder:latest] - Container log output (if any) will follow:
12:01:26 12:01:26.561 [dockerjava-netty-1-13] INFO  docker[richnorth/vnc-recorder:latest] - STDERR: Traceback (most recent call last):
12:01:26 12:01:26.561 [dockerjava-netty-1-13] INFO  docker[richnorth/vnc-recorder:latest] - STDERR: File "/usr/local/bin/flvrec.py", line 125, in <module>
12:01:26 12:01:26.561 [dockerjava-netty-1-13] INFO  docker[richnorth/vnc-recorder:latest] - STDERR: if __name__ == "__main__": sys.exit(main(sys.argv))
12:01:26 12:01:26.561 [dockerjava-netty-1-13] INFO  docker[richnorth/vnc-recorder:latest] - STDERR: File "/usr/local/bin/flvrec.py", line 123, in main
12:01:26 12:01:26.561 [dockerjava-netty-1-13] INFO  docker[richnorth/vnc-recorder:latest] - STDERR: debug=debug, verbose=verbose)
12:01:26 12:01:26.561 [dockerjava-netty-1-13] INFO  docker[richnorth/vnc-recorder:latest] - STDERR: File "/usr/local/bin/flvrec.py", line 24, in flvrec
12:01:26 12:01:26.561 [dockerjava-netty-1-13] INFO  docker[richnorth/vnc-recorder:latest] - STDERR: pwdcache = PWDFile(pwdfile)
12:01:26 12:01:26.561 [dockerjava-netty-1-13] INFO  docker[richnorth/vnc-recorder:latest] - STDERR: File "/usr/local/lib/python2.7/dist-packages/vnc2flv/rfb.py", line 50, in __init__
12:01:26 12:01:26.561 [dockerjava-netty-1-13] INFO  docker[richnorth/vnc-recorder:latest] - STDERR: fp = file(fname)
12:01:26 12:01:26.562 [dockerjava-netty-1-13] INFO  docker[richnorth/vnc-recorder:latest] - STDERR: IOError: [Errno 2] No such file or directory: '/recording/password'
12:01:26 12:01:26.562 [main] ERROR docker[selenium/standalone-firefox-debug:2.53.1] - Could not start container
12:01:26 org.testcontainers.containers.ContainerLaunchException: Container startup failed

**Comment by rnorth** [comment]
@cchanning sorry for not responding sooner. This seems a little odd but it seems like a volume mounting issue, probably due to the sibling container run mode. Still, I know of a team using Jenkins in (I think) a similar configuration so it should be possible.

By any chance is the job execution container using a volume mapping that makes the working directory resolve to a different directory on the host? (E.g. If you ran `pwd` inside of your build job and got `/foo`, is that directory something different on the host, such as `/bar/foo`?

That would seem like the most obvious way that volume mounting files into the containers could become broken.

**Comment by srempfer** [comment]
@cchanning: I had the same problem with Docker4Windows and Windows 10. After some investigations I found out that my anti virus software / firewall blocked the bind mount.

**Comment by dgouvea** [comment]
Maybe this log can help:

```
[main] ERROR 🐳 [richnorth/vnc-recorder:latest] - Container log output (if any) will follow:
[dockerjava-netty-1-11] INFO 🐳 [richnorth/vnc-recorder:latest] - STDERR: Traceback (most recent call last):
[dockerjava-netty-1-11] INFO 🐳 [richnorth/vnc-recorder:latest] - STDERR: File "/usr/local/bin/flvrec.py", line 125, in <module>
[dockerjava-netty-1-11] INFO 🐳 [richnorth/vnc-recorder:latest] - STDERR: if __name__ == "__main__": sys.exit(main(sys.argv))
[dockerjava-netty-1-11] INFO 🐳 [richnorth/vnc-recorder:latest] - STDERR: File "/usr/local/bin/flvrec.py", line 123, in main
[dockerjava-netty-1-11] INFO 🐳 [richnorth/vnc-recorder:latest] - STDERR: debug=debug, verbose=verbose)
[dockerjava-netty-1-11] INFO 🐳 [richnorth/vnc-recorder:latest] - STDERR: File "/usr/local/bin/flvrec.py", line 24, in flvrec
[dockerjava-netty-1-11] INFO 🐳 [richnorth/vnc-recorder:latest] - STDERR: pwdcache = PWDFile(pwdfile)
[dockerjava-netty-1-11] INFO 🐳 [richnorth/vnc-recorder:latest] - STDERR: File "/usr/local/lib/python2.7/dist-packages/vnc2flv/rfb.py", line 50, in __init__
[dockerjava-netty-1-11] INFO 🐳 [richnorth/vnc-recorder:latest] - STDERR: fp = file(fname)
[dockerjava-netty-1-11] INFO 🐳 [richnorth/vnc-recorder:latest] - STDERR: IOError: [Errno 2] No such file or directory: '/recording/password'
```

**Comment by dgouvea** [comment]
@rnorth I had a similar problem, I am running Jenkins as a Docker container. When I try to add some file as volume of a container, it's a problem, because the file is saved (let's say) in Jenkins workspace, but when you send the reference for Docker, it'll find the in the host, not in the Docker container.
In order to fix that, I've re-create the Jenkins container, sharing the /tmp from the host as a volume in the container. Now when I create some temporary file/folder, it's in the save path for Jenkins and Docker (host).
But it does not apply for the VncRecording, because it's creating a directory in the current path (GenericContainer.createVolumeDirectory) not in /tmp.
```
this.tempDir = this.createVolumeDirectory(true);
```
That's way the volume is not created correctly.

**Comment by kiview** [comment]
@dgouvea Have you tried with Testcontainers 1.5.0? The new version will create the recordings in the temp dir.

**Comment by dgouvea** [comment]
Thank you so much @kiview !

Now the container starts, but I still have an exception:

```
java.lang.RuntimeException: Could not move recording file from .tmp-volume-1513334990835/screen.flv to /tmp/containers/video/FAILED-ContainerTest-null-20171215-105332.flv
	at org.testcontainers.containers.VncRecordingSidekickContainer.stopAndRetainRecording(VncRecordingSidekickContainer.java:88)
	at org.testcontainers.containers.BrowserWebDriverContainer.stopAndRetainRecordingForDescriptionAndSuccessState(BrowserWebDriverContainer.java:228)
	at org.testcontainers.containers.BrowserWebDriverContainer.failed(BrowserWebDriverContainer.java:199)
	at org.testcontainers.containers.FailureDetectingExternalResource$1.evaluate(FailureDetectingExternalResource.java:34)
	at org.testcontainers.containers.FailureDetectingExternalResource$1.evaluate(FailureDetectingExternalResource.java:30)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
Caused by: java.nio.file.NoSuchFileException: .tmp-volume-1513334990835/screen.flv
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixCopyFile.move(UnixCopyFile.java:409)
	at sun.nio.fs.UnixFileSystemProvider.move(UnixFileSystemProvider.java:262)
	at java.nio.file.Files.move(Files.java:1395)
	at org.testcontainers.containers.VncRecordingSidekickContainer.stopAndRetainRecording(VncRecordingSidekickContainer.java:86)
	... 18 more
```

**Comment by dgouvea** [comment]
Another thing is that the BrowserWebDriverContainer is not too extensible, so I cannot override the creation of the VncRecordingContainer.

```
VncRecordingSidekickContainer recordingSidekickContainer = new VncRecordingSidekickContainer(this);
recordingSidekickContainer.start();
this.currentVncRecordings.add(recordingSidekickContainer);
```

The `this.currentVncRecordings` is private, so even I override the `containerIsStarted` method, I cannot add a different recording container.

Maybe another way to allow it is extract the creation of the VncContainer to another protected mehtod:

```
VncRecordingSidekickContainer recordingSidekickContainer = createVncRecordingContainer();
recordingSidekickContainer.start();
this.currentVncRecordings.add(recordingSidekickContainer);
...
protected VncRecordingSidekickContainer createVncRecordingContainer() {
    return new VncRecordingSidekickContainer(this);
}
```

In this case we could extend the `VncRecordingSidekickContainer` and override the `createVolumeDirectory`.

Of course, it's just an idea, I don't know if it could fit in the framework goals.

**Comment by dgouvea** [comment]
It works!

```
@ClassRule
    public static BrowserWebDriverContainer chrome = new BrowserWebDriverContainer(){
        @Override
        protected void containerIsStarted(InspectContainerResponse containerInfo) {
            super.containerIsStarted(containerInfo);

            withRecordingMode(VncRecordingMode.RECORD_ALL, new File("/tmp/containers/video"));
            VncRecordingSidekickContainer recordingSidekickContainer = new VncRecordingSidekickContainer(this){
                @Override
                protected Path createVolumeDirectory(boolean temporary) {
                    Path directory = (new File("/tmp/tmp-volume-" + System.currentTimeMillis())).toPath();
                    PathUtils.mkdirp(directory);
                    if(temporary) {
                        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
                            PathUtils.recursiveDeleteDir(directory);
                        }));
                    }

                    return directory;
                }
            };
            recordingSidekickContainer.start();

            try {
                Field currentVncRecordingsField = BrowserWebDriverContainer.class.getDeclaredField("currentVncRecordings");
                ReflectionUtils.makeAccessible(currentVncRecordingsField);
                Collection<VncRecordingSidekickContainer> currentVncRecordings = (Collection<VncRecordingSidekickContainer>) ReflectionUtils.getField(currentVncRecordingsField, this);
                currentVncRecordings.add(recordingSidekickContainer);
            } catch (Throwable e) {
                logger().error("Cannot change currentVncRecordings", e);
            }
        }
    }.withDesiredCapabilities(DesiredCapabilities.chrome())
            .withRecordingMode(BrowserWebDriverContainer.VncRecordingMode.SKIP, null);
```

**Comment by bsideup** [comment]
@dgouvea 

FYI I'm working on making VNC recorder attachable to any container with VNC enabled (Selenium's included) thanks to Networks support. 

I was demoing it at Heisenbug Moscow 2017, went well :) Will share the details soon (it kinda fixes this issue so I'll also report here)

**Comment by dgouvea** [comment]
Looks great @bsideup !

Thank you for the information!

**Comment by DaZang** [comment]
@bsideup 
Sounds great. Any updates on this?

**Comment by bsideup** [comment]
@DaZang sure! Released almost 1 month ago as 1.5.1 (FYI the latest is 1.6.0) :) Check `VncRecordingContainer`

**Comment by kiview** [comment]
I'll close this for now, since the problems should be fixed with `VncRecordingContainer`. If there are other errors with the new approach, we should create a new issue for them.
